doc/developer/design/20231102_read_holds.md

```diff
@@ -0,0 +1,156 @@
+For each object in the `HOLD`, the time must be in advance of that object's read frontier *or* any other `HOLD`'s time on that object, otherwise an error occurs.
+We want to support safe swapping out of one `HOLD` for another, and since we only reference a collection's `implied_capability` (which is not aware of any read holds),
+it is necessary to call out specifically that other `HOLD`s times are valid for new `HOLD`s.
+If `AT <time>` is not specified, Materialize automatically chooses the least valid read frontier among all objects of the `HOLD`.
+
+`HOLD`s are destroyed by:
+
+```
+DROP HOLD <name>
+```
+
+A `DROP` will release its current read hold.
+In-progress `SUBSCRIBE`s and transactions will not be affected after a `DROP`, but will continue to run as expected.
+Those will still be aquiring their own read holds in addition to the `HOLD` object.
+
+Creating a `HOLD` will increase resource consumption (mostly memory) in proportion to the amount of time it is held for.
+Users must periodically advance the `HOLD` to a more recent time to allow for resources to be released.
+
+`HOLD`s are advanced by:
+
+```
+ALTER HOLD <name> ADVANCE [TO <time>]
+```
+
+If `TO <time>` is specified, the new time will be created at that time.
+It is an error if that time is less than any object's logical read frontier or less than the existing `HOLD`'s time.
+If `TO <time>` is not specified, the new time is the least common physical read frontier of all objects named in the `HOLD`.
+The `HOLD` releases its current read hold and aquires a new read hold at the new time.
+
+The `MAX LAG` option (required, but defaults to `'3h'`) is a duration.
+Its upper bound is limited by a LD-configured setting.
+If difference between the write frontier (TODO: maybe another frontier?) of the objects and the `HOLD` time is greater than `MAX LAG`, the `HOLD`'s time is automatically advanced to `write frontier - MAX HOLD`.
+(Transactions and `SUBSCRIBE` are not affected, as in the `ALTER HOLD ADVANCE` case.)
+`SELECT` and `SUBSCRIBE` `AS OF` a time that is before a `HOLD`'s time will return a helpful error message that a `HOLD` exists but is in advance of the `AS OF` time.
+
+It is an error to `DROP` an object that is part of a `HOLD` unless the `DROP` is `CASCADE`.
+If the `DROP` is `CASCADE`, any `HOLD`s on the dropped object are also dropped (even if the `HOLD` is also over other objects).
+
+`HOLD`s cannot be mutated (objects added or removed) after creation, but they can be renamed in a transaction like other objects, and so can achieve the same goal.
+
+In order to specify a `HOLD` on a new index or materialized view whose underlying object may have a long history,
+`HOLD`s support a special multi-statement DDL of the form
+
+```
+BEGIN;
+CREATE INDEX i ...;
+CREATE READ HOLD ON i;
+COMMIT;
+```
+
+In a transaction, exactly one `CREATE HOLD` statement is allowed to follow a `CREATE` statement if the `HOLD` names exactly and only the created object.
```

Comment:
I might kick this to future work? It doesn't seem like we need to block releasing holds on the existence of multi-statement DDL, as long as we have a clear path to supporting this flow.

Reply:
Agree. I'm leaving in the text but will prioritize appropriately.